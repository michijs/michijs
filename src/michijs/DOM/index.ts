export { render, renderSync } from "./render";
export { create } from "./create/create";
