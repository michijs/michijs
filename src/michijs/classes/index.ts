export { ProxiedValue } from "./ProxiedValue";
export { Observable } from "./Observable";
export { CookieStorage } from "./CookieStorage";
export { EventDispatcher } from "./EventDispatcher";
export { ObservableFromEventListener } from "./ObservableFromEventListener";
export { HistoryManager } from "./HistoryManager";
export { VirtualFragment } from "./VirtualFragment";
export { MappedIdGenerator } from "./MappedIdGenerator";
export { IdGenerator } from "./IdGenerator";
export { ProxiedArray } from "./ProxiedArray";
