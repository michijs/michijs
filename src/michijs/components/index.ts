export { ElementInternals } from "./ElementInternals";
export { Host } from "./Host";
export { If } from "./If";
export { Router } from "./Router";
export { Fragment } from "./Fragment";
export { Title } from "./Title";
export { Redirect } from "./Redirect";
export { AsyncComponent } from "./AsyncComponent";
export { Slot } from "./Slot";
export { List } from "./List";
