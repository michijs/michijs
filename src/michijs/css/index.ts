export { cloneStylesheet } from "./cloneStylesheet";
export { useStyleSheet, cssObjectToText } from "./useStyleSheet";
export { useCssVariables } from "./useCssVariables";
export { useAnimation } from "./useAnimation";
export { useTransition } from "./useTransition";
export { css } from "./css";
export { convertCssObjectToCssVariablesObject } from "./convertCssObjectToCssVariablesObject";
