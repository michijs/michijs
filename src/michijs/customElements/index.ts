export { customElement } from "./customElement";
export { createElementProperties } from "./createElementProperties";
export { createCustomElement } from "./createCustomElement";
