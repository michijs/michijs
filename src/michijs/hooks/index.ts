export { useObserve } from "./useObserve";
export { useComputedObserve } from "./useComputedObserve";
export { useFetch } from "./useFetch";
export { useHash } from "./useHash";
export { usePromise } from "./usePromise";
export { useStorage } from "./useStorage";
export { usePureFunction } from "./usePureFunction";
export { useIndexedDB } from "./useIndexedDB";
export { useSearchParams } from "./useSearchParams";
export { useStringTemplate } from "./useStringTemplate";
export { useTitle } from "./useTitle";
export { useWatch } from "./useWatch";
export { useAsyncComputedObserve } from "./useAsyncComputedObserve";
