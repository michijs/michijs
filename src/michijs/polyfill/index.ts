import { getBrowser } from "../utils/getBrowser";

const isSafari = ["safari", "iphone"].includes(getBrowser());

let createBuiltInElement: typeof window.customElements.define;
if (isSafari) {
  const { overrideDocumentCreateElement, safariDefine } = await import(
    "./safariBuiltInElements"
  );
  createBuiltInElement = safariDefine;
  overrideDocumentCreateElement();
} else
  createBuiltInElement = window.customElements.define.bind(
    window.customElements,
  );

export { createBuiltInElement };
