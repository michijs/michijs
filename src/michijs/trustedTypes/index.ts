const returnParameter = (x: string): string => x;

const trustedTypeObject: TrustedTypePolicyOptions = {
  createHTML: returnParameter,
  createScript: returnParameter,
  createScriptURL: returnParameter,
};

export const trustedTypePolicy: TrustedTypePolicy =
  window.trustedTypes
    ?
    window.trustedTypes.createPolicy("michijs", trustedTypeObject)
    : (trustedTypeObject as unknown as TrustedTypePolicy);

export const makeMichijsTheDefaultTrustedPolicy = () => {
  if (window.trustedTypes)
    window.trustedTypes.createPolicy("default", trustedTypeObject)
}
