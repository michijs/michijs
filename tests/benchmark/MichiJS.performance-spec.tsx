import { type Browser, chromium, type Page } from "playwright-core";
import { installPlaywright, makePerformanceTests } from "./shared";
import { describe, it, expect, beforeEach, afterAll } from "bun:test";
import { spawn } from "child_process";
import { omit } from "@michijs/michijs";
import { writeFileSync } from "fs";
import michijs from "./results/michijs.json";
import packagejson from "../../package.json";
import { updateDiff } from "./updateDiff";
const serverProcess = spawn("bun", ["run", "start"], {
  stdio: "inherit",
  env: { ...process.env, NODE_ENV: "TESTING" },
});
await installPlaywright();
describe("Performance tests - MichiJS", () => {
  let browser: Browser;
  let page: Page;
  beforeEach(async () => {
    browser = await chromium.launch({
      headless: true,
    });
    page = await browser.newPage();
    await page.goto("http://localhost:3000", {
      waitUntil: "domcontentloaded",
    });
  });

  it("matches the snapshot on initial render", async () => {
    const body = await page.$("body");
    const innerHTMLProperty = await body!.getProperty("innerHTML");
    const innerHTML = await innerHTMLProperty.jsonValue();
    expect(innerHTML).toMatchSnapshot();
  });

  const resultsPromise = makePerformanceTests(
    () => browser,
    () => page,
  );
  afterAll(async () => {
    const results = await resultsPromise;
    const resultsString = JSON.stringify(
      {
        [packagejson.version]: results,
        ...omit(michijs, [packagejson.version]),
      },
      undefined,
      2,
    );
    writeFileSync("./tests/benchmark/results/michijs.json", resultsString);
    console.log("Results: ", JSON.stringify(results, undefined, 2));
    updateDiff(results);
    serverProcess.kill(2);
    browser.close();
  });
});

// Personal notes about performance:
// appendItem forEach (with referenced function) > appendItem for in > append forEach > fragment for in
// for in > forEach Object.keys when has no inherited enumerable properties
// for of > forEach
// Do not create unnecesary callbacks - move them to a separated function
// childList[i] > chilList.item(i)
