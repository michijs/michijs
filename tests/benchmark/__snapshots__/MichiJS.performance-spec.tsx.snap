// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Performance tests - MichiJS matches the snapshot on initial render 1`] = `
"
    <div id="main">
        <div class="container">
            <div class="jumbotron">
                <div class="row">
                    <div class="col-md-6">
                        <h1>MichiJS keyed</h1>
                    </div>
                    <div id="table-manager" class="col-md-6">
                    <div class="row"><div class="col-sm-6 smallpad"><button type="button" class="btn btn-primary btn-block" id="run">Create 1,000 rows</button></div><div class="col-sm-6 smallpad"><button type="button" class="btn btn-primary btn-block" id="runlots">Create 10,000 rows</button></div><div class="col-sm-6 smallpad"><button type="button" class="btn btn-primary btn-block" id="add">Append 1,000 rows</button></div><div class="col-sm-6 smallpad"><button type="button" class="btn btn-primary btn-block" id="update">Update every 10th row</button></div><div class="col-sm-6 smallpad"><button type="button" class="btn btn-primary btn-block" id="clear">Clear</button></div><div class="col-sm-6 smallpad"><button type="button" class="btn btn-primary btn-block" id="swaprows">Swap Rows</button></div></div></div>
                </div>
            </div>
            <table is="michi-table" class="table table-hover table-striped test-data">
            <tbody id="tbody"></tbody></table>
            <span class="preloadicon glyphicon glyphicon-remove" aria-hidden="true"></span>
        </div>
    </div>
    <script src="./index.js" type="module"></script>


"
`;
